### PR TITLE
Improve the performance of d2i_AutoPrivateKey and friends

### PR DESCRIPTION
**Pull request by mattcaswell**
Probe first to see if we have a PKCS8 file to improve decoder performance.

### CODE REVIEW

**Comment by mattcaswell** [comment]
I tested this using this code:

```` C
#include <string.h>
#include <openssl/evp.h>

static const char private_header[] = {
    0x30, 0x2e,          /* SEQUENCE(3 elem) (46 bytes)     */
    0x02, 0x01, 0x00,    /*   INTEGER 0                     */
    0x30, 0x05,          /*   SEQUENCE(1 elem) (5 bytes)    */
    0x06, 0x03,          /*     OBJECT IDENTIFIER (3 bytes) */
    0x2b, 0x65, 0x6e,    /*       OBJ_X25519 -> 1.3.101.110 -> 1*40+3, 101, 110 */
    0x04, 0x22,          /*   OCTET STRING(1 elem) (nested) */
    0x04, 0x20           /*     OCTET STRING(32 bytes)      */
};

static int dopkcs8_decode(const unsigned char *privdata, size_t privdata_len)
{
    int ret = 0;
    unsigned char der[64];
    const unsigned char *pder = der;
    long derlen;
    EVP_PKEY *key = NULL;
    size_t sz = sizeof(private_header);
    int selection;

    memcpy(der, private_header, sz);
    memcpy(der + sz, privdata, privdata_len);
    sz += privdata_len;
    derlen = sz;

    key = d2i_AutoPrivateKey(NULL, &pder, derlen);

    if (key == NULL)
        goto err;

    ret = 1;
 err:
    EVP_PKEY_free(key);
    return ret;
}

int main(void)
{
    unsigned int i, j;
    unsigned char private[32] = { 0 };

    for (i = 0; i < 256; i++) {
        private[0] = i;
        for (j = 0; j < 256; j++) {
            private[1] = j;
            if (!dopkcs8_decode(private, sizeof(private)))
                abort();
        }
    }
    return 1;
}
````

On master I see this:

````
real	0m15.269s
user	0m15.264s
sys	0m0.005s
````

On master with this patch I see:

````
real	0m8.859s
user	0m8.859s
sys	0m0.000s
````

**Comment by mattcaswell** [comment]
It seems to me we ought to be able to do this kind of probing in the decoder framework itself on an opportunistic basis for well known decoder chains. But there does not seem to be an obvious place to put it because the decoder API builds the decoder chain *before* any data is presented for actual decoding.

**Comment by richsalz** [comment]
One place to consider having "well-known chains" is when a decoder is registered (you've already locked things).

File: crypto/asn1/d2i_pr.c [hunk 1]
```diff
@@ -32,43 +33,55 @@ d2i_PrivateKey_decoder(int keytype, EVP_PKEY **a, const unsigned char **pp,
...
-            *pp = p;
-            goto err;
-        }
+            *a = *ppkey;
+        return *ppkey;
     }
-    /* Fall through to error if all decodes failed */
-err:
+
+ err:
```
Comment (levitte) [inline crypto/asn1/d2i_pr.c:84]: Side note: maybe we'd need to have something in the coding style regarding leftmost labels?  I see this space being randomly added or dropped, depending on who does it (supposedly depending on how the particular editor treats indentation, or perhaps running indent, or something like that)...
(Indenting leftmost labels with one space is a GNU style thing, as far as I recall, and they do have a reasoning around it, but I haven't read it for a long time, so don't remember)

**Comment by levitte** [comment]
> It seems to me we ought to be able to do this kind of probing in the decoder framework itself on an opportunistic basis for well known decoder chains. But there does not seem to be an obvious place to put it because the decoder API builds the decoder chain _before_ any data is presented for actual decoding.

Yes, and this is kind of a catch-22.  How do you detect the contents before having done some amount of parsing?  Or in other words, how much of the decoder's job do you want to replicate?

I might remind you, BTW, we do have content detection in some of the decoder, where it made sense to have it.


The (EncryptedPrivateKeyInfo) DER → (PrivateKeyInfo) DER decoder, which incidently also handles unencrypted input (i.e. a PrivateKeyInfo):

https://github.com/openssl/openssl/blob/d0a3b9d1eb1fc510ec3447b44803bbf5520a0c47/providers/implementations/encode_decode/decode_epki2pki.c#L135-L138

The PEM -> DER decoder does something similar too:

https://github.com/openssl/openssl/blob/d0a3b9d1eb1fc510ec3447b44803bbf5520a0c47/providers/implementations/encode_decode/decode_pem2der.c#L103-L135
https://github.com/openssl/openssl/blob/d0a3b9d1eb1fc510ec3447b44803bbf5520a0c47/providers/implementations/encode_decode/decode_pem2der.c#L175-L190

**Comment by mattcaswell** [comment]
> The (EncryptedPrivateKeyInfo) DER → (PrivateKeyInfo) DER decoder, which incidently also handles unencrypted input (i.e. a PrivateKeyInfo):

Yes, I know about that. I was actually playing around with that yesterday. I think it doesn't kick in though if you explicitly ask for the "PrivateKeyInfo" structure. I wondered about having a PrivateKeyInfo -> PrivateKeyInfo decoder (which just detected the type) and even prototyped something. But I wasn't seeing any speed up unfortunately.

**Comment by levitte** [comment]
> I think it doesn't kick in though if you explicitly ask for the "PrivateKeyInfo" structure.

Er...  what?  That sounds unfortunate, and is potentially a bug.

Or oh, aha, I think I understand what's going on, it gets side-stepped if the structure of the input is already known to be PrivateKeyInfo...  something that I guess also happens when the original input is a PEM file with a "PRIVATE KEY" header.  Hmmm, we should really fix that.

File: crypto/asn1/d2i_pr.c (hunk 1 above)
Comment (mattcaswell) [reply crypto/asn1/d2i_pr.c:84]: Interesting. I just did a quick regex to check relative numbers of the two forms in the current codebase.

Number of labels with exactly one space: 1782
Number of labels with no space: 734

So it seems like the clear preference in the current codebase is to have one space. That said, looking in the code style document, nothing explicit is said about this but the one code example in there which uses a label has no spaces in front of it.

**Comment by mattcaswell** [comment]
> I guess also happens when the original input is a PEM file with a "PRIVATE KEY" header.

Haven't checked that case - but possibly.

It also happens in the AutoPrivateKey code which explicitly states either "PrivateKeyInfo" or "type-specific". I guess it can also happen if people call the decoder code directly and specify "PrivateKeyInfo" explicitly (which seems quite reasonable if they know their private key is not encrypted).

File: crypto/asn1/d2i_pr.c (hunk 1 above)
Comment (levitte) [reply crypto/asn1/d2i_pr.c:84]: I think that the main reason why we have a "stronger preference" for the one space indent is the code reformatting we did years ago.  If I'm not mistaken, this is the indent parameter that regulates this:

https://github.com/openssl/openssl/blob/d0a3b9d1eb1fc510ec3447b44803bbf5520a0c47/util/indent.pro#L15

Comment (paulidale) [reply crypto/asn1/d2i_pr.c:84]: My personal preference is no space.
However, I thought our style was a single space.

**Comment by mattcaswell** [comment]
Ping?

**Review by beldmit — APPROVED** [review]
LGTM

File: crypto/asn1/d2i_pr.c (hunk 1 above)
Comment (levitte) [reply crypto/asn1/d2i_pr.c:84]: It is what we got with the indent configuration, but I don't think we have it written down in our coding style...

**Comment by mattcaswell** [comment]
So...just checking: This PR was created before 3.1 was final, but didn't make it in before the code freeze and therefore isn't part of 3.1.0. It currently has the "branch: 3.1" label on it. But seeing as this is a performance fix it doesn't meet the stable release updates policy.

Should I apply this just to master?

**Comment by richsalz** [comment]
The poor performance of 3.0 should really be considered a bug.  Just like previously it considered "missing accessors/settors" a bug when structures were made opaque.  Of course improvements that require huge changes (for some value of huge) may not make it.  But self-contained things that work to bring 3.0 performance closer to 1.1 really should go in.

**Comment by beldmit** [comment]
If I remember, as 3.1 includes performance fixes, we mentioned an option to consider similar issues as bugs for 3.1.

If necessary, I can put a hold on this to discuss it separately.

**Comment by t8m** [comment]
I think this should go into 3.1 branch as one of its objectives were to fix 3.0 performance issues. So in that sense it is a bug fix on 3.1 branch.

File: crypto/asn1/d2i_pr.c (hunk 1 above)
Comment (kaduk) [reply crypto/asn1/d2i_pr.c:84]: zero spaces would confuse the heuristic `git diff` uses to tell what function the changed lines are in

**Comment by paulidale** [comment]
Merged.

**Comment by beldmit** [comment]
@paulidale to both branches?

**Comment by mattcaswell** [comment]
> @paulidale to both branches?

Yes - looks like it went to both branches. That may have been premature given the discussion above - I'm not sure we had yet reached consensus.

Even though this is now post-merge, I've reopened the PR and added an OTC hold to it.

OTC Question: Is it ok that this was merged to 3.1?

**Comment by t8m** [comment]
**OTC: it is OK to keep it merged in 3.1 branch.**

**Pull request closed without merging**